### PR TITLE
Feat/add apify integration

### PR DESCRIPTION
**Pull request by jirispilka**
### Description

Implement functionality to retrieve Apify datasets and call Apify actors/tasks. 

See slack thread for [details](https://apify.slack.com/archives/C04URRT3934/p1721402814315259)

This integration is heavily inspired by the LangChain integration, specifically [apify_dataset.py](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/document_loaders/apify_dataset.py) and [apify_wrapper.py](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/utilities/apify.py)

Key differences:
- In LangChain, the `ApifyWrapper` class has a method to call actors with parameters like `actor_id`, `run_input`, and `dataset_mapping_function` cf. [call actor](https://github.com/langchain-ai/langchain/blob/4fab8996cf3a5a34bd5333c6848b0bccf798a6a0/libs/community/langchain_community/utilities/apify.py#L55).
- In Haystack, these parameters are moved to the class constructor. This change is made for the following reasons:
  - Passing arguments to a function in the [Haystack pipeline](https://docs.haystack.deepset.ai/docs/pipelines) is cumbersome (basically, one needs to pass dict to a function). It is more convenient to create a class instance with predefined parameters, add it as a pipeline component, and then call the `run` function.
  - Haystack users might prefer to run actors similarly to how it is done in the UI: select an actor, define input, and run.

CI/CD was copied from [crawlee](https://github.com/apify/crawlee-python/tree/master)

Plan:
- Add unit tests
- Add a key-value store loader 

### Testing

Manual

### Checklist

- [x] Changes are described in the `CHANGELOG.md`
- [x] CI passed

### CODE REVIEW

**Comment by jirispilka** [comment]
Kubo, I apologize for the extra commits and the noise I have created in your inbox 🙄

File: src/apify_haystack/__init__.py [hunk 1]
```diff
@@ -1,3 +1,7 @@
 from importlib import metadata
```
Comment (drobnikj) [inline src/apify_haystack/__init__.py:1]: nit: Why the source code is under apify_haystack folder, can it be directly in src?

File: README.md [hunk 2]
```diff
@@ -1,24 +1,68 @@
...
+
+In the example below, specify `apify_api_token` and run the script:
+
+```python
+from dotenv import load_dotenv
+from haystack import Document
+
+from apify_haystack import ApifyDatasetFromActorCall
+
+# Sey APIFY-API-TOKEN here or load it from .env file
```
Comment (drobnikj) [inline README.md]:
```suggestion
# Set APIFY_API_TOKEN here or load it from .env file
```

File: src/apify_haystack/examples/apify_actor_call.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+Haystack is an open-source framework fo...', meta: {'url': 'https://docs.haystack.deepset.ai/docs/intro'}
+.....
+"""
+
+from dotenv import load_dotenv
+from haystack import Document
+
+from apify_haystack import ApifyDatasetFromActorCall
+
+# Set APIFY-API-TOKEN here or load it from .env file
```
Comment (drobnikj) [inline src/apify_haystack/examples/apify_actor_call.py]:
```suggestion
# Set APIFY_API_TOKEN here or load it from .env file
```

File: src/apify_haystack/examples/crawl_and_process_data.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+from haystack.components.preprocessors import DocumentCleaner, DocumentSplitter
+from haystack.components.writers import DocumentWriter
+from haystack.document_stores.in_memory import InMemoryDocumentStore
+from haystack.utils.auth import Secret
+
+from apify_haystack import ApifyDatasetFromActorCall
+
+load_dotenv()
+
+# Set APIFY-API-TOKEN here or use it from .env file
```
Comment (drobnikj) [inline src/apify_haystack/examples/crawl_and_process_data.py]:
```suggestion
# Set APIFY_API_TOKEN here or use it from .env file
```

File: src/apify_haystack/examples/rag_with_crawled_website.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+from haystack.components.builders import PromptBuilder
+from haystack.components.embedders import OpenAIDocumentEmbedder, OpenAITextEmbedder
+from haystack.components.generators import OpenAIGenerator
+from haystack.components.retrievers.in_memory import InMemoryEmbeddingRetriever
+from haystack.document_stores.in_memory import InMemoryDocumentStore
+from haystack.utils.auth import Secret
+
+from apify_haystack import ApifyDatasetFromActorCall
+
+# Set APIFY-API-TOKEN here or use it from .env file
```
Comment (drobnikj) [inline src/apify_haystack/examples/rag_with_crawled_website.py]:
```suggestion
# Set APIFY_API_TOKEN here or use it from .env file
```

File: README.md [hunk 6]
```diff
@@ -1,24 +1,68 @@
 # Apify-Haystack integration
 
-Apify-Haystack integration allows to quickly fetch results from Apify Actors.
+The Apify-Haystack integration allows easy interaction between the Apify platform and Haystack.
```
Comment (drobnikj) [inline README.md]: Let's link Apify website and Haystack here.

File: README.md [hunk 7]
```diff
@@ -1,24 +1,68 @@
...
+Apify-haystack is available at the [`apify-haystack`](https://pypi.org/project/apify-haystack/) PyPI package.
 
 ```sh
 pip install apify-haystack
 ```
 
 ## Examples
 
-Here are some practical examples to help you get started ...
+Crawl website using Apify Website Content Crawler and convert it to Haystack Documents.
```
Comment (drobnikj) [inline README.md]:
```suggestion
### Crawl a website Using Apify Website Content Crawler and convert it to Haystack Documents
```

File: README.md [hunk 8]
```diff
@@ -1,24 +1,68 @@
...
+print("You can monitor the progress at: https://console.apify.com/actors/runs")
+
+dataset = actor.run().get("documents")
+
+print(f"Loaded {len(dataset)} documents from the Apify Actor {actor_id}:")
+for d in dataset:
+    print(d)
+```
+
+See other examples in the [examples directory](https://github.com/apify/apify-haystack/blob/master/src/apify_haystack/examples) for more examples, here is a list of few of them
```
Comment (drobnikj) [inline README.md:62]:
```suggestion
### More Examples

See other examples in the [examples directory](https://github.com/apify/apify-haystack/blob/master/src/apify_haystack/examples) for more examples, here is a list of few of them
```

**Review by drobnikj — CHANGES_REQUESTED** [review]
Code-wise, it looks good, please add something from the content team to check the readme and other texts before releasing.

File: src/apify_haystack/__init__.py (hunk 1 above)
Comment (jirispilka) [reply src/apify_haystack/__init__.py:1]: Good question. I followed the [scr layout](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/), e.g. `crawlee-python` is using that too.

File: README.md (hunk 2 above)
Comment (jirispilka) [reply README.md]: Good catch. Thanks!

File: src/apify_haystack/__init__.py (hunk 1 above)
Comment (drobnikj) [reply src/apify_haystack/__init__.py:1]: ok, I don't know that it is a common way in Python, so it is fine then.

**Review by drobnikj — APPROVED** [review]

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,6 +1,13 @@
 # Changelog
 
-## [0.0.1](https://github.com/apify/apify-haystack/releases/tag/v0.0.1) (2024-07-25)
+## [0.1.0] Not yet released  (2024-07-30)
 
+🚀 Features
+- Added support for Apify-Haystack integration
```
Comment (BraniganLee) [inline CHANGELOG.md]:
```suggestion
- Added support for the Apify-Haystack integration
```

File: README.md [hunk 10]
```diff
@@ -1,24 +1,70 @@
...
+Apify-haystack is available at the [`apify-haystack`](https://pypi.org/project/apify-haystack/) PyPI package.
 
 ```sh
 pip install apify-haystack
 ```
 
 ## Examples
 
-Here are some practical examples to help you get started ...
+### Crawl a website Using Apify Website Content Crawler and convert it to Haystack Documents
```
Comment (BraniganLee) [inline README.md]:
```suggestion
### Crawl a website using Apify's Website Content Crawler and convert it to Haystack Documents
```

File: README.md [hunk 11]
```diff
@@ -1,24 +1,70 @@
...
+print("You can monitor the progress at: https://console.apify.com/actors/runs")
+
+dataset = actor.run().get("documents")
+
+print(f"Loaded {len(dataset)} documents from the Apify Actor {actor_id}:")
+for d in dataset:
+    print(d)
+```
+
+### More Examples
```
Comment (BraniganLee) [inline README.md]:
```suggestion
### More examples
```

File: README.md [hunk 12]
```diff
@@ -1,24 +1,70 @@
...
+
+See other examples in the [examples directory](https://github.com/apify/apify-haystack/blob/master/src/apify_haystack/examples) for more examples, here is a list of few of them
+
+- Load a dataset from Apify and convert it to Haystack Document
```
Comment (BraniganLee) [inline README.md]:
```suggestion
- Load a dataset from Apify and convert it to a Haystack Document
```

File: README.md [hunk 13]
```diff
@@ -1,24 +1,70 @@
...
+- Call [Website Content Crawler](https://apify.com/apify/website-content-crawler) Actor and convert the data into the Haystack Documents
```
Comment (BraniganLee) [inline README.md]:
```suggestion
- Call [Website Content Crawler](https://apify.com/apify/website-content-crawler) and convert the data into the Haystack Documents
```

File: README.md [hunk 14]
```diff
@@ -1,24 +1,70 @@
...
+- Crawl website, retrieve text content, and store it in the `InMemoryDocumentStore`
```
Comment (BraniganLee) [inline README.md]:
```suggestion
- Crawl websites, retrieve text content, and store it in the `InMemoryDocumentStore`
```

File: scripts/update_version_for_prerelease.py [hunk 15]
```diff
@@ -24,12 +24,17 @@
...
+    else:
+        raise RuntimeError(f"Unknown prerelease type: {prerelease_type}")
 
     current_version = get_current_package_version()
 
     # We can only transform a stable release version (X.Y.Z) to a prerelease version (X.Y.ZxxxN)
     if not re.match(r"^\d+\.\d+\.\d+$", current_version):
-        raise RuntimeError(f"The current version {current_version} does not match the proper semver format for stable releases (X.Y.Z)")
+        raise RuntimeError(
+            f"The current version {current_version} does not match the proper semver format for "
```
Comment (BraniganLee) [inline scripts/update_version_for_prerelease.py:35]:
```suggestion
            f"The current version {current_version} does not match the proper server format for "
```

File: src/apify_haystack/apify_dataset.py [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+from apify_client import ApifyClient
+from haystack import component
+from haystack.dataclasses import Document
+
+from apify_haystack.constants import HAYSTACK_ATTRIBUTE_USER_AGENT
+
+
+@component
+class ApifyDatasetLoader:
+    """Load datasets produced by the `Apify Actors`.
```
Comment (BraniganLee) [inline src/apify_haystack/apify_dataset.py]:
```suggestion
    """Load datasets produced by `Apify Actors`.
```

File: src/apify_haystack/apify_dataset.py [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+        Type-hint note: the output is not type-hinted, otherwise linting complains
+        that `run` method is not haystack a component
+        """
+        dataset_items = self.client.dataset(self.dataset_id).list_items(clean=True).items
+        return {"documents": list(map(self.dataset_mapping_function, dataset_items))}
+
+
+@component
+class ApifyDatasetFromActorCall:
+    """Get Apify dataset by calling Apify Actor.
```
Comment (BraniganLee) [inline src/apify_haystack/apify_dataset.py]:
```suggestion
    """Get an Apify dataset by calling an Apify Actor.
```

File: src/apify_haystack/apify_dataset.py [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+
+    Perform call of Apify actor and load produced dataset.
```
Comment (BraniganLee) [inline src/apify_haystack/apify_dataset.py]:
```suggestion
    Perform call of Apify Actor and load produced dataset.
```

File: src/apify_haystack/apify_dataset.py [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+    For details, see https://docs.apify.com/platform/integrations/haystack
+
+    To use, you should have the ``apify-client`` python package installed,
```
Comment (BraniganLee) [inline src/apify_haystack/apify_dataset.py:72]: To use what?

Consider: To use it,

File: src/apify_haystack/apify_dataset.py [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+            dataset_mapping_function=self.dataset_mapping_function,
+        )
+        return loader.run()
+
+
+@component
+class ApifyDatasetFromTaskCall:
+    """Get Apify dataset by calling Apify Task.
+
+    Perform call of Apify task and load produced dataset.
```
Comment (BraniganLee) [inline src/apify_haystack/apify_dataset.py]:
```suggestion
    Perform a call of an Apify task and load the produced dataset.
```

File: src/apify_haystack/apify_dataset.py [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+        loader = ApifyDatasetLoader(
+            dataset_id=dataset_id,
+            dataset_mapping_function=self.dataset_mapping_function,
+        )
+        return loader.run()
+
+
+@component
+class ApifyDatasetFromTaskCall:
+    """Get Apify dataset by calling Apify Task.
```
Comment (BraniganLee) [inline src/apify_haystack/apify_dataset.py]:
```suggestion
    """Get an Apify dataset by calling an Apify task.
```

File: src/apify_haystack/apify_dataset.py [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+    ) -> None:
+        """Initialize the Apify Actor loader.
+
+        Args:
+            task_id (str): The ID or name of the Actor on the Apify platform.
+            task_input (Dict): The input object of the Actor that you're trying to run.
+            dataset_mapping_function (Callable): A function that takes a single
+                dictionary (an Apify dataset item) and converts it to an
+                instance of the Document class.
+            build (str, optional): Optionally specifies the actor build to run.
```
Comment (BraniganLee) [inline src/apify_haystack/apify_dataset.py]:
```suggestion
            build (str, optional): Optionally specifies the Actor build to run.
```

File: src/apify_haystack/examples/apify_actor_call.py [hunk 23]
```diff
@@ -0,0 +1,49 @@
+"""
+Call Apify Actor and load a dataset to convert it to Haystack Documents.
```
Comment (BraniganLee) [inline src/apify_haystack/examples/apify_actor_call.py]:
```suggestion
Call an Apify Actor and load a dataset to convert it to Haystack Documents.
```

File: src/apify_haystack/examples/apify_actor_call.py [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+
+This script demonstrates how to use Apify's Website Content Crawler to scrape text content from a specified website
+and convert the scraped data into Haystack Documents.
+
+For example, if you want to crawl and get text content from https://haystack.deepset.ai/, you can use this script
+to call the Apify actor and retrieve the data in a structured format.
```
Comment (BraniganLee) [inline src/apify_haystack/examples/apify_actor_call.py]:
```suggestion
to call the Apify Actor and retrieve the data in a structured format.
```

File: src/apify_haystack/examples/apify_dataset_load.py [hunk 25]
```diff
@@ -0,0 +1,31 @@
+"""
+Load a dataset from Apify and convert it to Haystack Documents.
+
+This script demonstrates how to load a dataset from Apify and convert it into Haystack Documents.
+Assuming you have run an Apify actor that scraped web pages from https://docs.haystack.deepset.ai/,
```
Comment (BraniganLee) [inline src/apify_haystack/examples/apify_dataset_load.py]:
```suggestion
Assuming you have run an Apify Actor that scraped web pages from https://docs.haystack.deepset.ai/,
```

File: src/apify_haystack/examples/crawl_and_process_data.py [hunk 26]
```diff
@@ -0,0 +1,89 @@
+"""
+Crawl website, scrape text content, and store it in the InMemoryDocumentStore.
```
Comment (BraniganLee) [inline src/apify_haystack/examples/crawl_and_process_data.py]:
```suggestion
Crawl websites, scrape text content, and store it in the InMemoryDocumentStore.
```

File: src/apify_haystack/examples/crawl_and_process_data.py [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+
+This script demonstrates how to extract content from a website using the Apify actor: Website Content Crawler.
```
Comment (BraniganLee) [inline src/apify_haystack/examples/crawl_and_process_data.py]:
```suggestion
This script demonstrates how to extract content from a website using Apify's Website Content Crawler.
```

**Review by BraniganLee — CHANGES_REQUESTED** [review]

File: scripts/update_version_for_prerelease.py (hunk 15 above)
Comment (jirispilka) [reply scripts/update_version_for_prerelease.py:35]: > 

Here, the semver stands for semantic version, so I'm skipping this suggestion

**Pull request merged**